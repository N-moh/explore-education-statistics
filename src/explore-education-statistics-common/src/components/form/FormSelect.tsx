import { Dictionary } from '@common/types';
import classNames from 'classnames';
import orderBy from 'lodash/orderBy';
import React, { ChangeEventHandler, FocusEventHandler, ReactNode } from 'react';
import ErrorMessage from '../ErrorMessage';
import styles from './FormSelect.module.scss';

export type SelectChangeEventHandler = ChangeEventHandler<HTMLSelectElement>;

export interface SelectOption {
  label: string;
  value: string | number;
  selected?: boolean;
}

export interface FormSelectProps {
  error?: string;
  id: string;
  label: ReactNode | string;
  name: string;
  onBlur?: FocusEventHandler;
  onChange?: SelectChangeEventHandler;
  options?: SelectOption[];
  optGroups?: Dictionary<SelectOption[]>;
  order?:
    | (keyof SelectOption)[]
    | ((option: SelectOption) => SelectOption[keyof SelectOption])[];
  orderDirection?: ('asc' | 'desc')[];
  value?: string | number;
}

const FormSelect = ({
  error,
  id,
  label,
  name,
  onBlur,
  onChange,
  options,
  optGroups,
  order = ['label'],
  orderDirection = ['asc'],
  value,
}: FormSelectProps) => {
  return (
    <>
      <label className="govuk-label" htmlFor={id}>
        {label}
      </label>
      {error && <ErrorMessage>{error}</ErrorMessage>}
      <select
        className={classNames('govuk-select', styles.select, {
          'govuk-select--error': !!error,
        })}
        id={id}
        name={name}
        onBlur={onBlur}
        onChange={onChange}
        value={value}
      >
        {options &&
          (order === undefined
            ? options
            : orderBy(options, order, orderDirection)
          ).map(option => (
            <option
              value={option.value}
              key={`${option.value}-${option.label}`}
              selected={option.selected}
            >
              {option.label}
            </option>
          ))}
        {optGroups &&
          Object.keys(optGroups).map(group => (
            <optgroup key={group} label={group}>
              {optGroups[group].map(option => (
                <option key={option.value} value={option.value}>
                  {option.label}
                </option>
              ))}
            </optgroup>
          ))}
      </select>
    </>
  );
};

export default FormSelect;
