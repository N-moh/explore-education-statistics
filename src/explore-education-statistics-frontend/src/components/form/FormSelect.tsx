import classNames from 'classnames';
import React, { ChangeEventHandler, ReactNode } from 'react';
import ErrorMessage from '../ErrorMessage';
import styles from './FormSelect.module.scss';

export interface SelectOption {
  text: string;
  value: string | number;
}

interface Props {
  error?: string;
  id: string;
  label: ReactNode | string;
  name: string;
  onChange?: ChangeEventHandler<any>;
  options: SelectOption[];
  value?: string | number;
}

const FormSelect = ({
  error,
  id,
  label,
  name,
  onChange,
  options,
  value,
}: Props) => {
  return (
    <>
      <label className="govuk-label" htmlFor={id}>
        {label}
      </label>
      {error && <ErrorMessage>{error}</ErrorMessage>}
      <select
        className={classNames('govuk-select', styles.select, {
          'govuk-select--error': !!error,
        })}
        id={id}
        name={name}
        onChange={onChange}
        value={value}
      >
        {options.map(option => (
          <option value={option.value} key={`${option.value}-${option.text}`}>
            {option.text}
          </option>
        ))}
      </select>
    </>
  );
};

export default FormSelect;
