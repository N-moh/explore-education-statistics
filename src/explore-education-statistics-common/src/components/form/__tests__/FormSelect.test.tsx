import React from 'react';
import { render } from 'react-testing-library';
import FormSelect from '../FormSelect';

describe('FormSelect', () => {
  test('renders correctly with required props', () => {
    const { container, getByLabelText } = render(
      <FormSelect
        id="test-select"
        label="Test select"
        name="testSelect"
        options={[
          { value: 'option-1', label: 'Option 1' },
          { value: 'option-2', label: 'Option 2' },
          { value: 'option-3', label: 'Option 3' },
        ]}
      />,
    );

    expect(getByLabelText('Test select')).toBeDefined();
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders list of options in alphabetical order by default', () => {
    const { getAllByText } = render(
      <FormSelect
        id="test-select"
        label="Test select"
        name="testSelect"
        options={[
          { value: 'option-2', label: 'Option 2' },
          { value: 'option-3', label: 'Option 3' },
          { value: 'option-1', label: 'Option 1' },
        ]}
      />,
    );

    const options = getAllByText(/Option/);

    expect(options[0]).toHaveTextContent('Option 1');
    expect(options[1]).toHaveTextContent('Option 2');
    expect(options[2]).toHaveTextContent('Option 3');
  });

  test('renders list of options in reverse alphabetical order', () => {
    const { getAllByText } = render(
      <FormSelect
        id="test-select"
        label="Test select"
        name="testSelect"
        orderDirection={['desc']}
        options={[
          { value: 'option-2', label: 'Option 2' },
          { value: 'option-3', label: 'Option 3' },
          { value: 'option-1', label: 'Option 1' },
        ]}
      />,
    );

    const options = getAllByText(/Option/);

    expect(options[0]).toHaveTextContent('Option 3');
    expect(options[1]).toHaveTextContent('Option 2');
    expect(options[2]).toHaveTextContent('Option 1');
  });

  test('renders list of options in custom order', () => {
    const { getAllByText } = render(
      <FormSelect
        id="test-select"
        label="Test select"
        name="testSelect"
        order={['value']}
        options={[
          { value: '1', label: 'Option 2' },
          { value: '2', label: 'Option 3' },
          { value: '3', label: 'Option 1' },
        ]}
      />,
    );

    const options = getAllByText(/Option/);

    expect(options[0]).toHaveTextContent('Option 2');
    expect(options[1]).toHaveTextContent('Option 3');
    expect(options[2]).toHaveTextContent('Option 1');
  });

  test('renders with error message', () => {
    const { container, getByText, getByLabelText } = render(
      <FormSelect
        id="test-select"
        error="Something went wrong"
        label="Test select"
        name="testSelect"
        options={[
          { value: 'option-1', label: 'Option 1' },
          { value: 'option-2', label: 'Option 2' },
          { value: 'option-3', label: 'Option 3' },
        ]}
      />,
    );

    expect(getByText('Something went wrong')).toHaveClass(
      'govuk-error-message',
    );
    expect(getByLabelText('Test select')).toHaveClass('govuk-select--error');
    expect(container.innerHTML).toMatchSnapshot();
  });
});
