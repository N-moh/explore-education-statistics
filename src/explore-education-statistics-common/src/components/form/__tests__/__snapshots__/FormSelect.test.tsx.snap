// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormSelect renders correctly with required props 1`] = `
<label class="govuk-label"
       for="test-select"
>
  Test select
</label>
<select class="govuk-select"
        id="test-select"
        name="testSelect"
>
  <option value="option-1">
    Option 1
  </option>
  <option value="option-2">
    Option 2
  </option>
  <option value="option-3">
    Option 3
  </option>
</select>
`;

exports[`FormSelect renders with error message 1`] = `
<label class="govuk-label"
       for="test-select"
>
  Test select
</label>
<span class="govuk-error-message">
  Something went wrong
</span>
<select class="govuk-select govuk-select--error"
        id="test-select"
        name="testSelect"
>
  <option value="option-1">
    Option 1
  </option>
  <option value="option-2">
    Option 2
  </option>
  <option value="option-3">
    Option 3
  </option>
</select>
`;
